xdsl/interpreters/experimental/wgpu.py

```diff
@@ -0,0 +1,214 @@
+        match (memref_type.element_type):
+            case IndexType():
+                element_size = 4
+            case _:
+                raise NotImplementedError(
+                    f"The element type {memref_type.element_type} for gpu.alloc is not implemented yet."
+                )
+        buffer = cast(
+            wgpu.GPUBuffer,
+            self.device.create_buffer(  # pyright: ignore
+                size=memref_type.element_count() * element_size,
+                usage=wgpu.BufferUsage.STORAGE  # pyright: ignore
+                | wgpu.BufferUsage.COPY_SRC,  # pyright: ignore
+            ),
+        )
+        return (buffer,)
+
+    @impl(gpu.MemcpyOp)
+    def run_memcpy(
+        self, interpreter: Interpreter, op: gpu.MemcpyOp, args: tuple[Any, ...]
+    ) -> tuple[()]:
+        """
+        Copy buffers according to the gpu.memcpy operation.
+
+        Only Device to Host copy is implemented here, to keep the first draft bearable.
+        """
+        src, dst = interpreter.get_values((op.src, op.dst))
+        if not (isinstance(src, wgpu.GPUBuffer) and isinstance(dst, ShapedArray)):
+            raise NotImplementedError(
+                f"Only device to host copy is implemented for now. got {src} to {dst}"
+            )
+
+        # Get device/source view
+        memview = cast(
+            memoryview, self.device.queue.read_buffer(src)  # pyright: ignore
+        )
+        dst_type = cast(MemRefType[Attribute], op.dst.type)
+        match (dst_type.element_type):
+            case IndexType():
+                format = "I"
+            case _:
+                raise NotImplementedError(
+                    f"copy for element type {dst_type.element_type} not yet implemented."
+                )
+        memview = memview.cast(format, [i.value.data for i in dst_type.shape])
+        for index in dst.indices():
+            dst.store(index, memview.__getitem__(index))  # pyright: ignore
+        return ()
+
+    @impl(gpu.LaunchFuncOp)
+    def run_launch_func(
+        self, interpreter: Interpreter, op: gpu.LaunchFuncOp, args: tuple[Any, ...]
+    ):
+        """
+        Launch a GPU kernel through the WebGPU API.
+        """
+        if op.asyncToken is not None or len(op.asyncDependencies) != 0:
+            raise NotImplementedError(
+                "The WGPU interpreter does not handle asynchronous GPU regions at the moment."
+            )
+
+        gridSize = interpreter.get_values((op.gridSizeX, op.gridSizeY, op.gridSizeZ))
+        blockSize = interpreter.get_values(
+            (op.blockSizeX, op.blockSizeY, op.blockSizeZ)
+        )
+        kernel_operands = op.kernelOperands
+
+        func = SymbolTable.lookup_symbol(op, op.kernel)
+        assert isinstance(func, gpu.FuncOp)
+        WGPUFunctions.compile_func(self, func)
+        shader_module = self.shader_modules[func]
+
+        # Compute the dispatch number
+        dispatch = list(gridSize)
```

Comment:
Looks like we don't really need the list construction, does this work?
```suggestion
        dispatch = gridSize
```

Reply:
Nope, tuple is immutable